feat(minifier): fold object constructor

camchenry opened the pull request:
This will fold expressions like `new Object()` to `{}`. Based on the closure compiler tests: https://github.com/google/closure-compiler/blob/b7e380b6320a52ce7fbac0334ede00d6a8c92fad/test/com/google/javascript/jscomp/PeepholeSubstituteAlternateSyntaxTest.java#L78.

This is outside my usual area, so feedback is welcome.

camchenry commented:
* **#6232** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6232?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @camchenry and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

Boshen commented on crates/oxc_minifier/src/ast_passes/peephole_substitute_alternate_syntax.rs:
```diff
@@ -260,6 +265,32 @@ impl<'a> PeepholeSubstituteAlternateSyntax {
+    fn compress_new_expression(&mut self, expr: &mut Expression<'a>, ctx: &mut TraverseCtx<'a>) {
+        if let Expression::NewExpression(new_expr) = expr {
+            // `new Object` -> `{}`
+            if new_expr.callee.is_global_reference_name("Object", ctx.symbols())
```
I would put the faster `is_empty` check first.

camchenry committed:
feat(minifier): fold object constructor (#6232)

This will fold expressions like `new Object()` to `{}`. Based on the closure compiler tests: https://github.com/google/closure-compiler/blob/b7e380b6320a52ce7fbac0334ede00d6a8c92fad/test/com/google/javascript/jscomp/PeepholeSubstituteAlternateSyntaxTest.java#L78.

This is outside my usual area, so feedback is welcome.
